src/analysis/convex/minmax.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2020 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import analysis.calculus.local_extr
+import topology.algebra.affine
+
+/-!
+# Minima and maxima of convex functions
+
+We show that if a function `f : E → ℝ` is convex, then a local minimum is also
+a global minimum.
+-/
+
+universes u
+
+variables {E : Type u} [add_comm_group E] [topological_space E]
+  [module ℝ E] [topological_add_group E] [topological_vector_space ℝ E]
+  {s : set E}
+
+open set filter
+open_locale classical
+
+/--
+Helper lemma for the more general case: `is_min_on.of_is_local_min_on_of_convex_on`.
+-/
+lemma is_min_on.of_is_local_min_on_of_convex_on_Icc {f : ℝ → ℝ} {a b : ℝ}
+  (a_lt_b : a < b) (h_local_min : is_local_min_on f (Icc a b) a) (h_conv : convex_on (Icc a b) f) :
+  ∀ x ∈ Icc a b, f a ≤ f x :=
+begin
+  by_contradiction H_cont,
+  push_neg at H_cont,
+  rcases H_cont with ⟨x, ⟨x_in_I, fx_lt_fa⟩⟩,
+
+  have exists_nhd_min : ∃ z > a, ∀ y ∈ (Icc a z), f y ≥ f a,
```

Comment:
`have` starting with an existential quantifiers are almost always better transformed into `obtain`. Here it would mean:
```lean
obtain ⟨z, hz, ge_on_nhd⟩ : ∃ z > a, ∀ y ∈ (Icc a z), f y ≥ f a,
```
and then get rid of the corresponding `rcases`.

Reply:
Yes, that looks much cleaner.